src/ui/win32/bindings/MemoryViewerControlBinding.cpp

```diff
@@ -290,6 +319,66 @@ void MemoryViewerControlBinding::OnViewModelIntValueChanged(const IntModelProper
     }
 }
 
+void MemoryViewerControlBinding::OnCopy()
+{
+    const auto& pEmulatorContext = ra::services::ServiceLocator::Get<ra::data::context::EmulatorContext>();
+    const auto nValue = pEmulatorContext.ReadMemory(m_pViewModel.GetAddress(), m_pViewModel.GetSize());
+    std::wstring sValue = ra::data::MemSizeFormat(nValue, m_pViewModel.GetSize(), MemFormat::Hex);
+
+    ra::services::ServiceLocator::Get<ra::services::IClipboard>().SetText(ra::Widen(sValue));
+}
+
+bool MemoryViewerControlBinding::OnPaste(bool bShiftHeld)
+{
+    auto nAddress = m_pViewModel.GetAddress();
+    const auto& pEmulatorContext = ra::services::ServiceLocator::Get<ra::data::context::EmulatorContext>();
+    std::wstring sClipboardText = ra::services::ServiceLocator::Get<ra::services::IClipboard>().GetText();
+    const auto nBytesForSize = ra::data::MemSizeBytes(m_pViewModel.GetSize());
+    std::vector<std::uint32_t> vValues;
+    std::wstring sError;
+
+    if (sClipboardText.empty())
+        return false;
+
+    // Split the clipboard value into substrings matching the current size and check if they're valid hexadecimal values
+    // If the clipboard text is smaller than a single write, it will be treated as a padded value:
+    //    C => 0C, 000C, or 0000000C (depending on currently selected viewer size)
+    // If shift is held, and the clipboard string is bigger than a single write, multiple writes will occur.
+    // If the last write is not a complete chunk, it will be padded in the same way as described above:
+    //    12345678C => 12 34 56 78 0C, 1234 5678 000C, or 12345678 0000000C
+    for (size_t i = 0; i < sClipboardText.size(); i += nBytesForSize * 2)
+    {
+        std::wstring sSubString = sClipboardText.substr(i, nBytesForSize * 2);
+        unsigned int nValue;
+
+        if (!ra::ParseHex(sSubString, 0xFFFFFFFF, nValue, sError))
+        {
+            ra::ui::viewmodels::MessageBoxViewModel::ShowWarningMessage(L"Paste value failed", sError);
+            return false;
+        }
+
+        // Single mode writes only the first value,
+        if (!bShiftHeld)
+        {
+            pEmulatorContext.WriteMemory(nAddress, m_pViewModel.GetSize(), nValue);
+            return true;
+        }
+
+        vValues.push_back(nValue);
```

Comment:
If you don't care about doing a full validation pass before writing, you can avoid using an array entirely and just write as you parse:
```
pEmulatorContext.WriteMemory(nAddress, m_pViewModel.GetSize(), nValue);
if (!bShiftHeld)
    return true;
nAddress += nBytesForSize;
```